src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -390,22 +386,67 @@ public void Dispose()
             /// <include file='doc\Queue.uex' path='docs/doc[@for="QueueEnumerator.MoveNext"]/*' />
             public bool MoveNext()
             {
-                if (_version != _q._version) throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
-
-                if (_index == -2)
-                    return false;
-
-                _index++;
-
-                if (_index == _q._size)
+                // Typically both of these conditions will be true, so save ourselves
+                // a branch and use bitwise AND.
+                bool typicalIteration = (_version == _q._version) & (_index != -2);
+
+                // Instead of return true/false, we assign the result to a variable and
+                // return it at the end. This helps decrease code size, as currently the
+                // jit cannot do this for us and generates code for 3 returns.
+                bool result = true;
+                if (typicalIteration)
                 {
-                    _index = -2;
-                    _currentElement = default(T);
-                    return false;
+                    _index++;
+
+                    if (_index == _q._size)
+                    {
+                        // We've run past the last element
+                        _index = -2;
+                        _currentElement = default(T);
+                        result = false;
+                    }
+                    else
+                    {
+                        // Cache some fields in locals to decrease code size
+                        T[] array = _q._array;
+                        int capacity = array.Length;
+
+                        // _index represents the 0-based index into the queue, however the queue
+                        // doesn't have to start from 0 and it may not even be stored contiguously in memory.
+
+                        int arrayIndex = _q._head + _index; // this is the actual index into the queue's backing array
+                        if (arrayIndex >= capacity)
+                        {
+                            // NOTE: Originally we were using the modulo operator here, however
+                            // on Intel processors it has a very high instruction latency which
+                            // was slowing down the loop quite a bit.
+                            // Replacing it with simple comparison/subtraction operations sped up
+                            // the average foreach loop by 2x.
+
+                            arrayIndex -= capacity; // wrap around if needed
+                        }
+                        
+                        _currentElement = array[arrayIndex];
+                    }
                 }
+                else
+                {
+                    result = MoveNextRare();
```

Comment:
@stephentoub MoveNext is not getting inlined. I only did this to decrease code size a bit since that code will rarely get used, so I didn't want to risk any possibility of it hurting the main path (e.g. spilling, longer jumps). [List](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Collections/Generic/List.cs#L1079) also does this, although `MoveNext` isn't inlined there either.